### ISSUE TITLE
Modelling GroupActions

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/LieGroups.jl#6]
I was wondering how to best model GroupActions.

Maybe

```
abstract type AbstractGroupAction{Group<:LieGroup, Manifold<:AbstractManifold, Direction <: AbstractGroupDirection}
```

(direction is basically just left and right.

But formerly (in Manifolds.jl) the apply also had a direction; is this double? Doe we then just need Covariate/Contravariate in apply? Or should we remove the direction from the group action?
In that case we could still let the direction in apply default to left.

We currently have `apply(action, g, p, dir)` (or even direction and co/contravariant).

Maybe Apply now just needs co/contra? 
For the docs afterwards I can help a bit, I would like to keep h,g for group elements, p,q for manifolds...to keep that a bit more consistent this time. But if someone wants to start this that would be great.

**Comment by olivierverdier** [comment]
> I was wondering how to best model GroupActions.
> 
> Maybe
> 
> ```
> abstract type AbstractGroupAction{Group<:LieGroup, Manifold<:AbstractManifold, Direction <: AbstractGroupDirection}
> ```

Why not, but I would keep the direction argument first.

> 
> (direction is basically just left and right.

To be clearer, the direction is whether the action is co/contravariant.

> But formerly (in Manifolds.jl) the apply also had a direction; is this double? 

It is. You can always do that by `apply(switch_direction(A), g, p)`.

> Doe we then just need Covariate/Contravariate in apply? Or should we remove the direction from the group action? In that case we could still let the direction in apply default to left.

I find the current implementation in `Manifolds.jl` satisfactory. Technically, you could have a direction argument in `apply` (that is co/contravariant), but I don't see how that would improve anything.

> 
> We currently have `apply(action, g, p, dir)` (or even direction and co/contravariant).

What do you mean by "even direction and co/contravariant"?

---

Just so that we are on the same page:

1. Any action can be co/contravariant. You can always switch between the two.
2. There are always two actions of a group on itself. We call them `LeftSide` and `RightSide`.

Combining 1. and 2., we see that each of those two actions of a group on itself can be either co or contravariant, just like any action. That gives us four actions on a group to itself.

And to be really super extra clear: `LeftSide` and `RightSide` **make no sense in general**, they only make sense when a group acts *on itself*.

And again, by contrast, co/contravariant **applies to all actions**.

**Comment by kellertuer** [comment]
> > I was wondering how to best model GroupActions.
> > Maybe
> > ```
> > abstract type AbstractGroupAction{Group<:LieGroup, Manifold<:AbstractManifold, Direction <: AbstractGroupDirection}
> > ```
> 
> Why not, but I would keep the direction argument first.

I would put the argument first that we use most, so if that is the direction, then sure. But having the other two in the abstract also already (they are not there in manifolds.jl) might be beneficial I think.

> > (direction is basically just left and right.
> 
> To be clearer, the direction is whether the action is co/contravariant.

Uff now you lost me again, after I maybe had finally thought I understood this. But maybe this all just stays magic to me.

For me direction just meant left -> left  group action; right -> right group action

and then indipendent of that (that is both are available for left/right) co/contra? And until now left(right for me was direction and co/contra did not yet have a name – now co/contra is direction – then what is left/right?

> > But formerly (in Manifolds.jl) the apply also had a direction; is this double?
> 
> It is. You can always do that by `apply(switch_direction(A), g, p)`.

But why is that double then? To confuse me? Also Direction for now always just was left/right – and somewhere some magic tuples appeared, how would these be stored in the GorupAction then?

> 
> > Doe we then just need Covariate/Contravariate in apply? Or should we remove the direction from the group action? In that case we could still let the direction in apply default to left.
> 
> I find the current implementation in `Manifolds.jl` satisfactory. Technically, you could have a direction argument in `apply` (that is co/contravariant), but I don't see how that would improve anything.
> 
> > We currently have `apply(action, g, p, dir)` (or even direction and co/contravariant).
> 
> What do you mean by "even direction and co/contravariant"?

Ah, I was wrong there, I though `apply` had as a last argument also the magic/direction parts. So then I am not sure where to put contra/co - but ou said left/right are contra/co anyways so I am not sure how many combinations we have here now. I first alwasys thought 4 then for a while 2 – now – no clue again. This seems to change twice a day.

> Just so that we are on the same page:
> 
> 1. Any action can be co/contravariant. You can always switch between the two.
> 2. There are always two actions of a group on itself. We call them `LeftSide` and `RightSide`.

We are very far from being on the same side. I had in mind any action is either a

* left action or a 
* right action

The rest – I have no clue, since that seems to change too often in all the posts. As soon as I have understood that I would probably document that as fast and as precise as possible.

> Combining 1. and 2., we see that each of those two actions of a group on itself can be either co or contravariant, just like any action. That gives us four actions on a group to itself.

where would that be stored? In direction? Should we rename direction? Somewhere else? how? Why? But ah so its 4 and at least not 8 16 or 32.
Because in my Head `AbstractActionDirection` (is the now first parameter of the action) would be either `LeftAction` or `RightAction` – namely whether we have g·p (group left, manifold right, · the acton, but we can also use another symbol) and p·g (group right)

How there other two are modelled (where I am aware there is a twist what is contra what is co, I have no clue eat about that switch) and where they would be stored, I have not yet understood. 


> And to be really super extra clear: `LeftSide` and `RightSide` **make no sense in general**, they only make sense when a group acts _on itself_.

to be extra clear – with this sentence I actively understand less even compared to what I ´thought I had understood before.

> And again, by contrast, co/contravariant **applies to all actions**.

See last comment. Is there any book that describes these 4 things with exactly this naming? Because Every time you write anting about these, in principle that always reads like an arbitrary permutation compared to everything you wrote before. I feel every time I read about the 4 it is different again and I understand less and less.

**Comment by olivierverdier** [comment]
> I would put the argument first that we use most, so if that is the direction, then sure. But having the other two in the abstract also already (they are not there in manifolds.jl) might be beneficial I think.

Excellent, I agree with that.

I see that you do not understand actions of a group on itself. I've explained this many times, we even found a better notation together: covariant/contravariant, and *for group actions on themselves*, there is also the left/right distinction.

I'll try to explain one last time. Take a deep breath, and please read slowly and carefully the following five paragraphs:

---

Actions always come in two flavours, covariant and contravariant. It's like head and tails for coins: all coins have a head and tail. Always.

For instance, a rotation action is such a coin. This coin has two sides, head and tail. But that is just **ONE** coin.

For a group acting on itself, there are **TWO** coins. Let me call these two coins "Yin" and "Yang".  Each with two sides. How many sides is that? That is four sides. You have "Yin, head", "Yin, tail", "Yang, head", "Yang, tail". 

Do not *ever* confuse "Yin" and "Yang", which are *coins*, with "head" and "tail", which are the sides that every coin has. 

You can't take a coin and say: "I want the Yin of that coin". That. does. not. make. sense.

---

Now please try to spend some time to understand this. Writing "now you lost me", "every time I understand less and less", "I have no clue", "This seems to change twice a day" makes it exceptionally challenging for me to help you understand better.

**Comment by kellertuer** [comment]
Hm, yeah sorry for my tone. My problem is: When you try to explain stuff, the tone is always “As we all obviously know, because if ou do not already know, you are stupid.”
If it is then later after a long day, maybe I am loosing a bit of productive things, especially if then there is now new names (and with ring and yang now even more names.

Here is my problem: Formerly we had a plethora of lefts and rights (left side, left action,...) – I never understood why.
Now I am reading Hilgert Neeb, and they write in Section 9.1.3, definition 9.1.11 and remark, that $\sigma(g,m)$ is a left action and that the other way around we have a right action. On the other hand you had here https://github.com/JuliaManifolds/LieGroups.jl/issues/3#issuecomment-2351010213 some left action and right action as well. There you also already said left and right are co- and contra variant. 

Now you write above that left and right do not make sense. So, instead of now coming with even more names and analogies – I would like to see a proper definition in your naming.
Because I feel that you are convinced that your naming is the only correct one and if we do not use it, no one will understand it (and I hope you do not mean it that strongly as it comes across) 

* What is a Left action?
* What is a Right action?
* What makes an action covariant?
* What makes an action contra variant?

In _as precise as possible_ mathematical definitions. No Ying, Yang, Coins, heads, tails, forward, backward, withershins, deosil,  obviously clear analogies, rim ward, hun ward,...bevause by now I am really lost on analogies. And that is also probably not your fault but mine.

If I do not understand the mathematical definitions, well then (or maybe even as a general idea) someone else implements and documents Actions and I take a huge step back and just care for a unified notation as I now started in https://github.com/JuliaManifolds/LieGroups.jl/blob/main/docs/src/notation.md or for usage in the docs (by interpolation like `$(_math(:G))` in https://github.com/JuliaManifolds/LieGroups.jl/blob/main/src/documentation_glossary.jl, where I try to collect usual terms and description snippets (e.g. for variable names and their description in fields/arguments/keyword arguments).
I will stick to Hilgert Neeb in literature because it seems that one I do understand, but any other is fine as well. Then I would like to carefully ask to find a common notation as well in the docs (but I can take the main responsibility for the docs, sure.)

In general I think I do not even _have_ to understand all this though I would like to, but by now I am convinced, that will not happen during this semester.

TLDR: I will now stop with mentioning that I do not understand what you write, but I would kindly ask to document every type, every function with a as precise as possible mathematical definition and for best of case a reference to some literature, where a terminology and/or a choice for a name is from.

**Comment by kellertuer** [comment]
I could still not find a good definition of this second variable idea (currenttly in Manifolds.jl –  Forward/Backward, ying/yang, left/right (confusing because we have that already), covariant/contravraiant, wither shins/deosil,...

So I read even more about this and tried to wrap my head around. Again, probably I am stupid. Then I thought – can‘t we just make it a bit easier? So I want to propose something.

I read Hilgert Neeb (2012) p. 291, and they introduce $\sigma_L:  \mathcal G \times \mathcal M \to \mathcal M$ as the group action (from the left) which they also shorten to just $\sigma$ – and use $g.m$ in short – I would maybe prefer $g\cdot m$, since in programming `.` means something different, and use that if it is clear from context which concrete group action $\sigma$ is meant.
Then they write in Remark 9.1.12 that one can also introduce a right action $\sigma_R: \mathcal M \times \mathcal M \to \mathcal M$ and if they are smooth we have $\sigma_L(g,m) = \sigma_R(m, g*^{-1})$ defines a left action as well. (and vice versa.

So. Why shall we invent own names that for now at least I can not find in literature and I do not understand where the are from? 
My proposal is

1. We do not call the abstract thing `AbstractGroupDirection ` but `AbstractGroupActonType`– this leaves us a bit more freedom
2. We introduce four such types
    * `LeftGroupAction` $\sigma_L(g,m)$
    * `RightGroupAction` $\sigma_R(m,g)$
    * `InverseLeftGroupAction` $\sigma_L^{-1}(g,m) = \sigma_L(g^{-1},m)$
    * `InverseRightGroupAction` $\sigma_R^{-1}(m,g)  = \sigma_R(m, g^{-1})$

Then these are
* easy to document
* they are actual inverses due to associativity group actions need to fulfil, that is $\sigma_L^{-1}(g, \sigma_L(g, m) = \sigma_L(g^{-1}, \sigma_L(g,m)) = \sigma_L(g^{-1} \circ g, m) = m$
* we get rid of all the direction-and-side or magic terms above discussion
* `switch_direction`, `switch`direction_and_side`,... would simplify to maybe less functions? For example the docs here https://github.com/JuliaManifolds/Manifolds.jl/blob/43b699062d9d3ad8fb1b39ad0805c1cbdb2da1e4/src/groups/group.jl#L190-L208 at least for me do not clarify what switching the side means (and both are called left/right which is confusing); the only thing I learn from the docs of `switch_side` is what it _not_ does.
* we could (simply) introduce `switch_action` (to switch between left action and right action) and `invert_action` (hopefully even clear from reading that this returns the inverse left for left vice versa), and we can carefully check where which of these and/or a combination of these is needed.
* Finally both switch have to really only be implemented to modify/generate a new Group action, and with the above relations are well-documentable and explainable what they actually switch/inverse. We also would not need two “switch” operations, but have clear and distinct names.

Furthermore with the action knowing its group and manifold, also `apply` hopefully also simplifies (with all its friends – inverse, diff,...).

Let me know what you think.

**Comment by mateuszbaran** [comment]
> My proposal is
> 
> 1. We do not call the abstract thing `AbstractGroupDirection ` but `AbstractGroupActonType`– this leaves us a bit more freedom
> 2. We introduce four such types
>    

That's fine for me. I don't want to use a notation here that isn't used by any textbook, even if it makes sense from a certain point of view.

> Furthermore with the action knowing its group and manifold, also `apply` hopefully also simplifies (with all its friends – inverse, diff,...).

I think we shouldn't need the `inverse_` functions if we have the inverse actions. All other points look reasonable.

**Comment by kellertuer** [comment]
Sure, `inverse_apply` we could skip

> > My proposal is
> > 
> > 1. We do not call the abstract thing `AbstractGroupDirection ` but `AbstractGroupActonType`– this leaves us a bit more freedom
> > 2. We introduce four such types
> 
> That's fine for me. I don't want to use a notation here that isn't used by any textbook, even if it makes sense from a certain point of view.

Nice.  That was also my main start for this proposal – to stick close to textbooks. I trued a few first steps to this proposal, and it seems reasonablz easy to realise.

> 
> > Furthermore with the action knowing its group and manifold, also `apply` hopefully also simplifies (with all its friends – inverse, diff,...).
> 
> I think we shouldn't need the `inverse_` functions if we have the inverse actions. All other points look reasonable.

right, we do not necessarily need those; we could keep them for ease-of-use – for example as

```
inverse_apply(A::AbstractGroupAction, g, p) = apply(inv(A), g, p)
```

but that is also not that necessary maybe. But let's also wait for Olivier here, because I feel he has a strong opinion on this. For me, the proposal I did is something I feel I understand, so that maybe others are also reasonably likely to understand what I document then.

**Comment by kellertuer** [comment]
I stumbled into an interesting naming question:

* In `ManifoldDiff` we name any differential as `differential_f` and sometimes `differential_v_f` for a function `f(v,w)`, sometimes referring to `v` with its name sometimes with its type (e.g. group/manifold)
* in the `GroupManifolds` area we usually have `f_diff_v` instead. 

Which shall we keep? I am leaning towards the first, new names make a few things in implementation here easier _and_ it's consistent throughout the packages then.
I checked, the only “area” that uses the second naming scheme is actually the group manifolds/group actions in `Manifolds.jl`.

**Comment by kellertuer** [comment]
Thanks for the discussion today, so I will rename all functions to the following scheme:

`modifier_indicator_f`
where
*  modifier might be `diff` or `inv` (for inverse, in front it is also easier to not be confused with inv_variant)
* indicator is used to indicate a variable, e.g. `group` or `manifold` in actions or `left` right` in compose. If there is just one thing we can inv/diff/... with the omit the indicator.
* and f is the function in question
